File: julia/src/pivot.jl
abstract type PivotRule end

"""
Pick the non-basic variable with the largest (most positive) reduced cost.
"""
struct LargestCoefficient <: PivotRule end

"""
Bland's rule: always pick the lowest-index positive reduced cost.
Guarantees finite termination independently of the ratio test, at the cost of
slower convergence on typical problems.
"""
struct Bland <: PivotRule end

function entering_index(::LargestCoefficient, cr::AbstractVector{Float64};
                         tol::Float64 = 1e-10) :: Union{Int, Nothing}
    v, j = findmax(cr)
    v > tol ? j : nothing
end

function entering_index(::Bland, cr::AbstractVector{Float64};
                         tol::Float64 = 1e-10) :: Union{Int, Nothing}
    findfirst(x -> x > tol, cr)
end

"""
Lexicographic ratio test — selects the leaving variable to prevent cycling.

Given:
- `xb`:    values of the current basic variables (length m)
- `d`:     entering column in basis coordinates, B⁻¹Aⱼ (length m)
- `BinvN`: full direction matrix B⁻¹N (m × |nonbasic|), used for tie-breaking

Returns the *row position* (into `basic`) of the leaving variable, or `nothing`
if every component of `d` is non-positive (problem is unbounded).

**Anti-cycling guarantee**: ties in the primary ratio `xb[i]/d[i]` are broken by
comparing rows of `BinvN / d[i]` lexicographically. Because `BinvN` is a matrix
with linearly independent columns from the LP's initial tableau, no two candidates
can tie on every column, ensuring a unique choice and finite termination.
"""
function leaving_index(xb::AbstractVector{Float64},
                        d::AbstractVector{Float64},
                        BinvN::AbstractMatrix{Float64};
                        tol::Float64 = 1e-10) :: Union{Int, Nothing}

    candidates = findall(i -> d[i] > tol, eachindex(d))
    isempty(candidates) && return nothing

    best = candidates[1]
    for i in @view candidates[2:end]
        ratio_i    = xb[i]    / d[i]
        ratio_best = xb[best] / d[best]

        if ratio_i < ratio_best - tol
            best = i
        elseif abs(ratio_i - ratio_best) <= tol
            # Lexicographic tie-break on scaled rows of BinvN
            row_i    = BinvN[i,    :] ./ d[i]
            row_best = BinvN[best, :] ./ d[best]
            Tuple(row_i) < Tuple(row_best) && (best = i)
        end
    end

    return best
end

File: typescript/src/pivot.ts
export type PivotRule = 'largest' | 'bland';

/**
 * Return the index into cr of the entering variable, or null if optimal.
 * Indices are 0-based (index into the nonbasic array).
 */
export function enteringIndex(
  rule: PivotRule,
  cr: number[],
  tol: number = 1e-10,
): number | null {
  if (rule === 'largest') {
    let best = -1;
    let bestVal = tol;
    for (let j = 0; j < cr.length; j++) {
      if (cr[j] > bestVal) { bestVal = cr[j]; best = j; }
    }
    return best === -1 ? null : best;
  } else {
    // Bland: first positive reduced cost
    for (let j = 0; j < cr.length; j++) {
      if (cr[j] > tol) return j;
    }
    return null;
  }
}

/**
 * Lexicographic ratio test — returns the row index (into basic) of the leaving
 * variable, or null if all d[i] <= tol (unbounded direction).
 *
 * @param xb    current basic variable values (length m)
 * @param d     entering column in basis coordinates B^{-1} A_j (length m)
 * @param BinvN full B^{-1} N matrix (m x |nonbasic|), used for tie-breaking
 */
export function leavingIndex(
  xb: number[],
  d: number[],
  BinvN: number[][],
  tol: number = 1e-10,
): number | null {
  const m = d.length;
  const candidates: number[] = [];
  for (let i = 0; i < m; i++) {
    if (d[i] > tol) candidates.push(i);
  }
  if (candidates.length === 0) return null;

  let best = candidates[0];
  for (let ci = 1; ci < candidates.length; ci++) {
    const i = candidates[ci];
    const ratioI = xb[i] / d[i];
    const ratioBest = xb[best] / d[best];

    if (ratioI < ratioBest - tol) {
      best = i;
    } else if (Math.abs(ratioI - ratioBest) <= tol) {
      // Lexicographic tie-break on scaled rows of BinvN
      const k = BinvN[0]?.length ?? 0;
      for (let col = 0; col < k; col++) {
        const rowI = BinvN[i][col] / d[i];
        const rowBest = BinvN[best][col] / d[best];
        if (rowI < rowBest - tol) { best = i; break; }
        if (rowI > rowBest + tol) { break; }
      }
    }
  }

  return best;
}
